parl/algorithms/paddle/cql.py

```diff
@@ -0,0 +1,295 @@
+#   Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import parl
+import paddle
+from paddle.distribution import Normal
+import paddle.nn.functional as F
+from parl.utils.utils import check_model_method
+from copy import deepcopy
+import numpy as np
+
+__all__ = ['CQL']
+
+
+class CQL(parl.Algorithm):
+    def __init__(self,
+                 model,
+                 gamma=None,
+                 tau=None,
+                 actor_lr=None,
+                 critic_lr=None,
+                 policy_eval_start=40000,
+                 with_automatic_entropy_tuning=True,
+                 with_lagrange=False,
+                 lagrange_thresh=10.0,
+                 min_q_version=3,
+                 min_q_weight=5.0,
+                 alpha=1.0):
+        """ CQL algorithm
+        Args:
+            model (parl.Model): forward network of actor and critic.
+            gamma (float): discounted factor for reward computation
+            tau (float): decay coefficient when updating the weights of self.target_model with self.model
+            actor_lr (float): learning rate of the actor model
+            critic_lr (float): learning rate of the critic model
+            policy_eval_start (int): try doing behaivoral cloning at the beginning, 40000 or 10000 work similarly
+            with_automatic_entropy_tuning (bool): train with automatic entropy tuning in Actor.
+            with_lagrange (bool): train with lagrange
+            lagrange_thresh (float): the value of tau, corresponds to the CQL(lagrange) version, suggest 10.0 in mujoco and 5.0 in Franka kitchen or Adroit domains
+            min_q_version (int): min_q_version = 3 (CQL(H)), = 2 (CQL(rho)), will be set to <0 in cql if not using lagrange
+            min_q_weight (float): the value of alpha in Critic loss, suggest 5.0 or 10.0 if not using lagrange
+            alpha (float): the value of alpha(temperature parameter) in Actor loss, determines the relative importance of entropy term against the reward
+        """
+        # checks
+        # check_model_method(model, 'value', self.__class__.__name__)
```

Comment:
Why do we comment these lines? 
I think they are necessary to reuse the algorithm?